### PR TITLE
[NavigationDrawer] Move off of MDCDeviceTopSafeAreaInset.

### LINKED ISSUE #4900: [NavigationDrawer] Move away from MDCDeviceTopSafeAreaInset
We should remove all `MDCDeviceTopSaveAreaInset` calls and change them to use safeAreaInsets as this is a soon to be deprecated API.

---
#### Internal data
- Associated internal bug: [b/117177500](http://b/117177500)

### PR DESCRIPTION
**Pull request by jakerockland**
Resolves #4900, which was closed due to inactivity, but is still a relevant issue.

We're currently running into an issue where, because `view.safeArea.topInset` is never used for non-notched devices, extra height is always added to our drawer headers on presentation to take account for the hard-coded status bar height of `20`, even when displayed in fullscreen landscape mode where this does not apply.

This is resolved by universally using the `safeArea` insets, rather than relying on `MDCDeviceTopSafeAreaInset`. This is done in favor of changing the behavior of `MDCDeviceTopSafeAreaInset` itself, as it is less clear to me what unintended effects that may have in other components.

For devices running on iOS prior to 11.0, this change is a no-op, as the changes in `MDCBottomDrawerViewController` and `MDCBottomDrawerContainerViewController` will just use the hardcoded value of `20` that is always returned by `MDCDeviceTopSafeAreaInset` in a pre-"iOS 11" context.

I have also removed usage of `addedContentHeightThreshold` as this was just a passthrough for accessing the top area inset and have consolidated to use one shared getter `topSafeAreaInset ` throughout the `MDCBottomDrawerContainerViewController`.

### CODE REVIEW

File: components/private/UIMetrics/src/MDCLayoutMetrics.h
Comment (jakerockland) [inline components/private/UIMetrics/src/MDCLayoutMetrics.h:24]: Exposing this constant in the header here hopefully will make it more straightforward for other components to move off of `MDCDeviceTopSafeAreaInset` in a similar way to how is being done the bottom drawer. However, if desired I'm happy to duplicate this constant within the bottom drawer and not modify `MDCLayoutMetrics`.

**Comment by jakerockland** [comment]
One additional note about this, it likely should not be merged until #9471 is merged, to prevent #8277 from beginning to repro in non-notched devices. There is more about this in my essay of a PR comment in #9471 😅

**Comment by jakerockland** [comment]
Exploring the open internal issues within this component, I realized that this PR likely also resolves #7371.

**Comment by yarneo** [comment]
This is great! Thank you!

**Review by yarneo — APPROVED** [review]

**Comment by yarneo** [comment]
I am changing your extern const to have an MDC prefix so not to collide with other apps. Also fixed your clang-format issues so we can land this by next release which is tomorrow

**Comment by jakerockland** [comment]
> I am changing your extern const to have an MDC prefix so not to collide with other apps. Also fixed your clang-format issues so we can land this by next release which is tomorrow

Ah awesome, thank you for doing so @yarneo, that is very much appreciated. 💯 😄  

I had thought the next release wasn't until Friday. Was I poorly pattern-matching thinking that MDC always cut on Fridays, or is that just some of the time? I can rebase + update #9464 tomorrow to include these changes as well as those in #9471. 

Given that there are not any unexpected changes in the behavior of those changes, do you think it would be possible that those changes then make it into a release before next Monday (02/03)? I am hoping to get them incorporated into some fixes before next Tuesday (02/04).

**Comment by yarneo** [comment]
Hey! I will try to land the 2 PRs into develop before the release. If you are able to test the sticky header PR before the release Monday then we can try to get that in as well!

In terms of release timing: it is up to the release cop for that week to decide when in the week to release (usually the earlier in the week the better). However specifically for the next 2 weeks we will be changing slightly the way we send PRs on GitHub and for us to integrate the change the release-cop will need to release early on Monday. Sorry for the inconvenience regarding this.

**Comment by yarneo** [comment]
Looks like there are some unit tests and snapshot failures, could you please take a look @jakerockland ?

**Comment by jakerockland** [comment]
Hey, not a worry at all! Thanks for the clarification on upcoming release schedules, that is super helpful.

If I have a chance tonight I will take a look at the broken tests as well as updating the “sticky header” PR to be rebased off of these changes. Otherwise we’ll get it done first thing tomorrow and we can plan on the following release.

Thanks again Yarden for taking the time to review.

**Comment by jakerockland** [comment]
Hey @yarneo! I've added a commit to this PR that hopefully makes this ready for the cut tomorrow, but I definitely need another set of eyes on this (more below).

Unfortunately I'm pretty uncertain why the snapshot test is failing. When I run the test suite from the `MDCCatalog.xcworkspace` project in the Product menu, the tests all pass for me. Likewise, when I run with blaze the tests all pass for me. Did the updated `testPresentedDrawerWithColoredViewsWithVerticalSizeClassCompact_11_2@2x.png` file need to be generated in a specific way that I may have missed? Not sure if this is something you can easily verify/update before the cut.

As far as the unit tests, it looks like when ran locally these run on a simulated device without safe area insets (why my updated tests passed for me) but fail when run on the runner targets `IPAD_PRO_12_9_IN_9_3` and `IPHONE_7_PLUS_IN_10_3` due to the safe areas being present. I don't _fully_ understand what all is going on in the build + test tooling being used here that makes this the case, but I've updated this PR to address for this _I think_. The tests still pass for me when ran locally with this update, but I'm having a hard time figuring out how I can verify these tests in the way they are performed by Kokoro. 😓

**Comment by yarneo** [comment]
Thanks for the update @jakerockland !
I'll put some time to try and see if I can easily resolve the issues on my end, but if not, we will probably have to wait for the next release. What we can do is try to cut another release later this week just for these changes if this is something urgent, but we are planning some changes to how we introduce changes to our repo, so we can chat about it more if we are unable to land it today

**Comment by romoore** [comment]
Running snapshots locally. Will push up the goldens if I can get them to update. Looks like Git LFS attributes on one of the images weren't set as expected, and that could have contributed to the failures.

**Comment by romoore** [comment]
I tried running these locally, but found the unit tests timing-out (not launching simulators). I'm not sure if it's changes in this PR or those specific targets, but I had no luck trying to get these tests to execute locally. Sorry!

**Comment by jakerockland** [comment]
@yarneo @romoore looking at the latest Kokoro run, it does look like the suspected device-specific safe area issue _was_ the issue, so that part has been fixed. From what I can tell it is just the snapshot issue that remains. If it is a Git LFS issue, it may be because I generated the golden with Blaze and then copied the exported image over? I'm not sure how to generate the goldens otherwise, so that they have the correct attributes.

**Comment by romoore** [comment]
I may have had an issue with my laptop. I was seeing Xcode hang when attempting to install to device as well.  I've rebooted and I can install to device.  Sorry for the delay.

This CL can be imported and landed internally as an internal change. It will then be exported back to GitHub in 2 weeks when we return to our typical GitHub-focused development workflow.  Does that work for you?  Internal screenshots are separate from those on GitHub, so the imported change will produce its own screenshot updates.

**Comment by jakerockland** [comment]
@romoore Not a worry at all! If we can get this change landed internally that would be fantastic and definitely works for us 🙌. Is there any chance that the change here could be moved internally and then back to Github in the same workflow #9464 (note that that PR was rebased off of this one, with the hopes of making it easier to get both in the pending release)? 

I have a handful of bugs currently that are blocked on the resolution of the two issues that these PRs address and am really hoping to be able to resolve them by the time we cut a release candidate again at the beginning of next week.

**Comment by romoore** [comment]
@jakerockland I know that @yarneo has been working with you on the two changes. I think he'll be best-suited to making sure the changes land internally (and back here) in a way that works well for your team.  Yarden - does that work for you, too?

**Comment by yarneo** [comment]
Yep, I'll be helping @jakerockland with this and the other PRs, thanks for taking a look

**Pull request closed without merging**
